README.md
# SMRF - Structure-based Markov random field
This is an evolutionary analysis tool using structure-based Markov random field.


## Installation

### Prerequisites
The standard Python package and the following packages are required.

- [Biopython]
- [Numpy]

The PMRF software (currently, compatible with v0.4.x) is required.

- [PMRF]

### Download
Download the the latest release from the [SMRF repository].

### Installation
This tool is not packaged for the general installation, yet. Instead, uncompress the downloaded file to the directory where the SMRF software will be installed.

### Configuring the Initial Setting
Create and configure smrf.cfg with the directory where the PMRF software is installed.

```sh
$ smrf.py   # This command creates smrf.cfg, if it does not exist in the directory where SMRF is installed
$ vi [SMRF_DIR]/smrf.cfg
```


## Usage

### Getting Started
Calculate the positional coevolution scores by using example.afa (MSA file formatted as FASTA) and example.pdb (PDB file).

```sh
$ smrf.py example.afa example.pdb
```

Show the help page.

```sh
$ smrf.py -h
```

### Writing coevolution scores
Write the positional and pairwise coevolution scores with `--pos` and `--pair` options, respectively.

```sh
$ smrf.py example.afa example.pdb --pos pos_score.txt --pair pair_score.txt
```

### Writing MRF model
Write the MRF model with `--mrf` option.

```sh
$ smrf.py example.afa example.pdb --mrf model.mrf
```

The MRF model is compatible with [PMRF] software.

## Reference
If you use this software for any published work, please cite the accompanying paper.

- Chan-Seok Jeong and Dongsup Kim (2016) Structure-based Markov random field model for representing evolutionary constraints on functional sites. BMC Bioinformatics. 17, 99.


[SMRF repository]: https://github.com/jeongchans/smrf/releases
[PMRF]: https://github.com/jeongchans/pmrf
[Biopython]: http://biopython.org/
[Numpy]: http://www.numpy.org/

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.


## 0.4.0 - 2016-06-13

### Changed
- Use PMRF v0.4.0, much faster and versatile.
- MRF file is written in PMRF v0.4.0 compatible format.

### Compatibility Warning
- Not support PMRF v0.2.x.


## 0.2.2 - 2016-06-02

### Added
- Specify the edge file for MRF architecture. Use `--edge` option.
- Specify the mrf file for SMRF model. Use `--mrf` option.
- Show the release version with `--version` option

### Changed
- Change the command-line options. Input files are required, but output files are optional.
- Require <msa_file> and <pdb_file>.
- Specify the output file for the positional coevolution scores. Use `--pos` option. If not specified, stdout will be used as default.
- Specify the output file for the pairwise coevolution scores. Use `--pair` option. If not specified, the pairwise coevoltion scores will not be reported.
- Show help message with `-h` and `--help` options.


## 0.2.1 - 2016-02-04

### Added
- This CHANGELOG file to contain notable changes.

### Changed
- Organize the installation procedure in README document.

### Fixed
- Fix the serious bug in VERSION assignment.
- Fix the initial generation of configuration file.
- Fix the link for the PMRF software in the README document.


## 0.2.0 - 2016-01-07 [YANKED]

### Added
- Build the structure-based Markov random field.
- Calculate the pairwise coevolution scores.
- Calculate the positional coevolution scores.
^

mrf.py
import subprocess
import os.path
import sys

import numpy as np

def run_pmrf(afa_file, edge_file, mrf_file, pmrf_path):
    opts = '%s --edge %s -o %s'%(afa_file, edge_file, mrf_file)
    pmrf_exec = '%s/pmrf'%pmrf_path
    if not os.path.exists(pmrf_exec):
        print 'Cannot find the PMRF executable in the directory %s.'%(pmrf_path)
        sys.exit(1)
    cmd = '%s build %s'%(pmrf_exec, opts)
    subprocess.check_call(cmd.split())

class MRFParser(object):

    def __init__(self, mrf_file, idx_beg):
        self._mrf_file = mrf_file
        self._idx_beg = idx_beg

    def _open_mrf_file(self):
        return open(self._mrf_file)

    def _parse_idx(self, x):
        return int(x) - 1 + self._idx_beg

    def iternodes(self):
        conv = lambda x: 0. if x == '*' else float(x)
        fp = self._open_mrf_file()
        for line in fp:
            if line.strip() == '# NODE': break
        fp.next()
        for line in fp:
            if line.strip() == '# EDGE': break
            cols = line.strip().split('\t')
            idx = self._parse_idx(cols[0].split()[1])
            v = np.array(map(conv, cols[1:]))
            yield MRFNode(idx, v)
        fp.close()

    def iteredges(self):
        conv = lambda x: 0. if x == '*' else float(x)
        fp = self._open_mrf_file()
        for line in fp:
            if line.strip() == '# EDGE': break
        fp.next()
        for line in fp:
            if line.strip() == '//': break
            cols = line.strip().split('\t')
            idx1 = self._parse_idx(cols[0].split()[1])
            idx2 = self._parse_idx(cols[1].split()[1])
            v = np.array(map(conv, cols[2:]))
            n = int(np.sqrt(len(v)))
            yield MRFEdge(idx1, idx2, v.reshape((n, n), order='C'))
        fp.close()

class MRFNode(object):

    def __init__(self, idx, weight):
        self.idx = idx
        self.weight = weight

class MRFEdge(object):

    def __init__(self, idx1, idx2, weight):
        self.idx1 = idx1
        self.idx2 = idx2
        self.weight = weight


edge.py
import itertools

from Bio import SeqIO
from Bio import pairwise2
import Bio.PDB.PDBParser

from util import open_file
from util import message

ONECODE = {'ALA':'A', 'CYS':'C', 'ASX':'B', 'GLU':'E', 'ASP':'D',
           'GLY':'G', 'PHE':'F', 'ILE':'I', 'HIS':'H', 'LYS':'K', 
           'XLE':'J', 'MET':'M', 'LEU':'L', 'PYL':'O', 'ASN':'N', 
           'GLN':'Q', 'PRO':'P', 'SER':'S', 'ARG':'R', 'SEL':'U', 
           'THR':'T', 'TRP':'W', 'VAL':'V', 'TYR':'Y', 'XAA':'X', 
           'GLX':'Z'}

class PDBParser(object):

    def parse_atom_seq(self, fp):
        residue_list = self.parse_residue_list(fp)
        return ''.join([ONECODE[x.get_resname().upper()] for x in residue_list])

    def parse_atom_cb(self, fp):
        residue_list = self.parse_residue_list(fp)
        ret = []
        for x in residue_list:
            if x.get_resname() == 'GLY': ret.append(self._get_atom(x, 'CA', None))
            elif x.has_id('CB'): ret.append(self._get_atom(x, 'CB', None))
            else: ret.append(self._get_atom(x, 'CA', None))
        return ret

    def parse_residue_list(self, fp):
        parser = Bio.PDB.PDBParser()
        structure = parser.get_structure('Read by PDBParser instance', fp)
        residue_list = structure.get_list()[0].get_list()[0].get_list()
        return [x for x in residue_list if x.get_id()[0].strip() == '']

    def _get_atom(self, res, k, d=None):
        return res[k] if res.has_id(k) else d

class IndexMapper(object):

    def __init__(self):
        self.MSK = 'X'

    def assign_index(self, seq, ref_seq, beg_idx=0):
        aln_seq, aln_ref_seq = pairwise2.align.globalms(seq.upper(), ref_seq.upper(), 1, -1, -.5, -.1)[0][:2]
        ref_idx_list = range(beg_idx, beg_idx + len(ref_seq))
        aln_ref_idx_list = self.build_aln_idx_list(ref_idx_list, aln_ref_seq)
        return self.assign_index_from_aln(aln_seq, aln_ref_seq, aln_ref_idx_list=aln_ref_idx_list)

    def build_aln_idx_list(self, idx_list, aln_seq):
        aln_idx_list, i = [], 0
        for y in aln_seq:
            if y != '-':
                aln_idx_list.append(idx_list[i])
                i += 1
            else: aln_idx_list.append(None)
        return aln_idx_list

    def assign_index_from_aln(self, aln_seq, aln_ref_seq, aln_ref_idx_list=[], msk_size=3, nonidentical=False):
        msk_aln_seq = self.mask_ambiguous_aln(aln_seq, msk_size)
        msk_aln_ref_seq = self.mask_ambiguous_aln(aln_ref_seq, msk_size)
        idx_list = []
        for x, y, curr in zip(msk_aln_seq, msk_aln_ref_seq, aln_ref_idx_list):
            if x != '-':
                if y != '-' and x != self.MSK and (nonidentical or x == y): idx_list.append(curr)
                else: idx_list.append(None)
        return idx_list

    def mask_ambiguous_aln(self, aln_seq, msk_size):
        frag_list = []
        for i, x in enumerate(aln_seq):
            if i == 0:
                frag_list.append(x)
            else:
                if prev_x == '-':
                    if x == '-': frag_list[-1] += x
                    else: frag_list.append(x)
                else:
                    if x == '-': frag_list.append(x)
                    else: frag_list[-1] += x
            prev_x = x
        frag_leng_list = []
        for frag in frag_list:
            if frag[0] == '-': frag_leng = 0
            else: frag_leng = len(frag)
            frag_leng_list += [frag_leng] * len(frag)
        ret = ''
        for x, n in zip(aln_seq, frag_leng_list):
            if x != '-' and n < msk_size: ret += self.MSK
            else: ret += x
        return ret

class Distance(object):

    def __init__(self, i, ri, j, rj, dist):
        self.i = i
        self.ri = ri
        self.j = j
        self.rj = rj
        self.dist = dist

_pdb_parser = PDBParser()
_idx_mapper = IndexMapper()

def build_edge(afa_file, pdb_file, cutoff=8.):
    seq = str(SeqIO.parse(open_file(afa_file), 'fasta').next().seq).upper().replace('-', '')
    message('Target sequence is retrieved from MSA.\n%s'%seq)
    atm_seq = _pdb_parser.parse_atom_seq(open_file(pdb_file)).upper()
    idx_list = _idx_mapper.assign_index(atm_seq, seq, beg_idx=1)
    if atm_seq != seq:
        message('Sequence in PDB is not identical to target sequence.\n%s'%atm_seq)
        message('Only the following residues will be considered for the analysis.\n' + ', '.join(['%s%d'%(res, i) for i, res in zip(idx_list, atm_seq) if i != None]))
    dist_list = _calc_dist(idx_list, atm_seq, pdb_file)
    edge_list = [(x.i, x.j) for x in dist_list if x.dist < cutoff]
    return edge_list

def _calc_dist(idx_list, atm_seq, pdb_file):
    dist_list = []
    cb_list = _pdb_parser.parse_atom_cb(open_file(pdb_file))
    for (i, res_i, cb_i), (j, res_j, cb_j) in itertools.combinations(zip(idx_list, atm_seq, cb_list), 2):
        if i == None or cb_i == None or j == None or cb_j == None: continue
        d = cb_i - cb_j
        if i > j: i, res_i, j, res_j = j, res_j, i, res_i
        dist_list.append(Distance(i, res_i, j, res_j, d))
    return dist_list


score.py
from Bio import SeqIO
import numpy as np

from util import open_file
from mrf import MRFParser

def calc_pos_score(mrf_file, afa_file):
    frac, psfm = _calc_psfm(afa_file)
    mrf_parser = MRFParser(mrf_file, idx_beg=1)
    score_list = [_calc_pos_score(node.weight[:-1], f) for node, f in zip(mrf_parser.iternodes(), psfm)]
    score_list = _rescale_score(score_list)
    vname_list = [node.idx for node in mrf_parser.iternodes()]
    return zip(vname_list, score_list)

def _rescale_score(scr_list):
    scr = np.array([x for x in scr_list if x != None])
    mn, sd = scr.mean(), scr.std()
    if sd > 0: sc = lambda x: (x - mn) / sd if x != None else None
    else: sc = lambda x: x - mn if x != None else None
    return [sc(x) for x in scr_list]

def _kl_divergence(p, q):
    return sum([x * np.log(x / y) for x, y in zip(p, q) if x > 0 and y > 0])

def _calc_pos_score(w, f):
    p = np.exp(w) / np.exp(w).sum()
    return _kl_divergence(f, p)

def _calc_psfm(afa_file):
    AA2Index = {aa:idx for aa, idx in zip('ACDEFGHIKLMNPQRSTVWY', range(20))}
    seqs = [str(r.seq) for r in SeqIO.parse(open_file(afa_file), 'fasta')]
    md_seq = lambda seq: ''.join([x for x, y in zip(seq, seqs[0]) if y != '-'])
    seqs = map(md_seq, seqs)
    n = len(seqs[0])
    f = np.zeros((n, 20), dtype=float)
    frac = np.zeros(n, dtype=float)
    for seq in seqs:
        for i, x in enumerate(seq):
            if AA2Index.has_key(x):
                if x == 'B': f[i, (AA2Index['N'], AA2Index['D'])] += .5
                elif x == 'J': f[i, (AA2Index['I'], AA2Index['L'])] += .5
                elif x == 'Z': f[i, (AA2Index['Q'], AA2Index['E'])] += .5
                elif x == 'O': f[i, AA2Index['K']] += 1.
                elif x == 'U': f[i, AA2Index['C']] += 1.
                else: f[i, AA2Index[x]] += 1.
                frac[i] += 1.
    f = f / np.reshape(np.repeat(f.sum(1), 20), (n, 20))
    frac /= len(seqs)
    return frac, f

def calc_pair_score(mrf_file):
    mrf_parser = MRFParser(mrf_file, idx_beg=1)
    score_list = _calc_pair_score(mrf_parser)
    zscore_list = _rescale_score([x.score for x in score_list])
    return [EdgeScore(x.i, x.j, zscore) for x, zscore in zip(score_list, zscore_list)]

def _calc_pair_score(mrf_parser):
    edge_score = lambda edge: EdgeScore(edge.idx1, edge.idx2, np.linalg.norm(edge.weight[:-1, :-1]))
    return _apc_correct([edge_score(edge) for edge in mrf_parser.iteredges()])

def _apc_correct(score_list):
    mn = {}
    for x in score_list:
        mn.setdefault(x.i, []).append(x.score)
        mn.setdefault(x.j, []).append(x.score)
    mn = {k:np.mean(v) for k, v in mn.iteritems()}
    tot_mn = np.mean([x.score for x in score_list])
    apc_func = lambda scr: scr.score - mn[scr.i] * mn[scr.j] / tot_mn
    ret = [EdgeScore(x.i, x.j, apc_func(x)) for x in score_list]
    return ret

class EdgeScore(object):

    def __init__(self, i, j, score):
        self.i = i
        self.j = j
        self.score = score


smrf.py
#!/usr/bin/env python
#
# smrf
#   Coevolution analysis tool by using structure-based Markov random field
#
# Written by Chan-Seok Jeong <csjeong@kaist.ac.kr>, 2014-2016.
#
# If you use this software for any published work, please cite the accompanying paper.
#
#   Chan-Seok Jeong, and Dongsup Kim (2016) Structure-based Markov random field model for representing evolutionary constraints on functional sites. BMC Bioinformatics. 17, 99.
#

import sys
import tempfile
import os.path
import optparse
import ConfigParser
import subprocess
import shlex

from edge import build_edge
from util import message
from util import write_file

VERSION = '0.4.0'
CONFIG_FILE = '%s/smrf.cfg'%os.path.dirname(__file__)
DEFAULT_CONFIG = """\
[external]
pmrf_path =                ; absolute path for PMRF software
"""

def load_config():
    if not os.path.exists(CONFIG_FILE):
        write_file(DEFAULT_CONFIG, CONFIG_FILE)
        print 'Configuration file is created as %s'%(CONFIG_FILE)
        raise SystemExit
    cfg = ConfigParser.ConfigParser()
    cfg.read(CONFIG_FILE)
    return {'pmrf_path': os.path.expanduser(cfg.get('external', 'pmrf_path').split(';')[0])}

def smrf(afa_file, pdb_file, options, pmrf_path):
    edge_file, mrf_file = options.edge_file, options.mrf_file
    pmrf_exec = '%s/pmrf'%pmrf_path
    if not os.path.exists(pmrf_exec):
        print 'Cannot find the PMRF executable in the directory %s.'%(pmrf_path)
        sys.exit(1)

    ## Determine graph structure
    edge_list = build_edge(afa_file, pdb_file)
    write_file('\n'.join(['%s\t%s'%(i, j) for i, j in edge_list]), edge_file)
    message('MRF edge is determined.')

    ## Build MRF model
    cmd = '%s build %s --edge %s -o %s'%(pmrf_exec, afa_file, edge_file, mrf_file)
    subprocess.check_call(cmd.split())
    message('MRF model is parameterized.')

    ## Estimate positional coevolution
    cmd = '%s stat %s --mode pos'%(pmrf_exec, mrf_file)
    fp = sys.stdout if options.score_file1 == "stdout" else open(options.score_file1, 'w')
    p = subprocess.Popen(shlex.split(cmd), stdout=fp)
    p.wait()
    message('Positional coevolution scores are estimated.')

    ## Estimate pairwise coevolution
    if options.score_file2:
        cmd = '%s stat %s --mode pair'%(pmrf_exec, mrf_file)
        p = subprocess.Popen(shlex.split(cmd), stdout=open(options.score_file2, 'w'))
        p.wait()
        message('Pairwise coevolution scores are estimated.')

def parse_options():
    parser = optparse.OptionParser(usage = "%prog <msa_file> <pdb_file> [options]",
                                   version="%prog " + VERSION)
    grp = optparse.OptionGroup(parser, "Output options")
    grp.add_option("--pos", dest="score_file1", default="stdout", help="write positional coevolution scores to <score_file> (default: stdout)", metavar="<score_file>")
    grp.add_option("--pair", dest="score_file2", default=None, help="write pairwise coevolution scores to <score_file>", metavar="<score_file>")
    grp.add_option("--edge", dest="edge_file", default=None, help="write MRF edge list to <edge_file>", metavar="<edge_file>")
    grp.add_option("--mrf", dest="mrf_file", default=None, help="write MRF model to <mrf_file>", metavar="<mrf_file>")
    parser.add_option_group(grp)
    (options, args) = parser.parse_args()
    if len(args) != 2: parser.error("Incorrect number of arguments. See '%s --help'."%os.path.basename(sys.argv[0]))
    for arg in args:
        if not os.path.exists(arg): parser.error("The file %s does not exist"%arg)
    (msa_file, pdb_file) = args
    return msa_file, pdb_file, options

if __name__ == '__main__':
    cfg = load_config()
    afa_file, pdb_file, options = parse_options()
    if not options.edge_file:
        edge_tf = tempfile.NamedTemporaryFile()
        options.edge_file = edge_tf.name
    if not options.mrf_file:
        mrf_tf = tempfile.NamedTemporaryFile()
        options.mrf_file = mrf_tf.name
    smrf(afa_file, pdb_file, options, cfg['pmrf_path'])
